File: templates/python/LocalPubSub/localpubsub/subscriber.py
import sys
import traceback

from awsiot.greengrasscoreipc.clientv2 import GreengrassCoreIPCClientV2
from awsiot.greengrasscoreipc.model import SubscriptionResponseMessage


def subscribe_to_topic(ipc_client: GreengrassCoreIPCClientV2, topic):
    return ipc_client.subscribe_to_topic(
        topic=topic, on_stream_event=_on_stream_event, on_stream_error=_on_stream_error, on_stream_closed=_on_stream_closed
    )


def _on_stream_event(event: SubscriptionResponseMessage) -> None:
    try:
        message = str(event.binary_message.message, "utf-8")
        topic = event.binary_message.context.topic
        print("Received new message on topic %s: %s" % (topic, message))
    except Exception as e:
        print("Exception occurred: " + str(e))
        traceback.print_exc()


def _on_stream_error(error: Exception) -> bool:
    print("Received a stream error.", file=sys.stderr)
    traceback.print_exc()
    return False  # Return True to close stream, False to keep stream open.


def _on_stream_closed() -> None:
    print("Subscribe to topic stream closed.")

File: templates/java/LocalPubSub/src/main/java/com/localpubsub/Subscriber.java
package com.localpubsub;

import software.amazon.awssdk.aws.greengrass.GreengrassCoreIPCClientV2;
import software.amazon.awssdk.aws.greengrass.model.BinaryMessage;
import software.amazon.awssdk.aws.greengrass.model.SubscribeToTopicRequest;
import software.amazon.awssdk.aws.greengrass.model.SubscriptionResponseMessage;

import java.nio.charset.StandardCharsets;
import java.util.Optional;

public class Subscriber {
    private final GreengrassCoreIPCClientV2 clientV2;

    public Subscriber(GreengrassCoreIPCClientV2 greengrassCoreIPCClientV2) {
        this.clientV2 = greengrassCoreIPCClientV2;

    }

    public GreengrassCoreIPCClientV2.StreamingResponse subscribe(String topic) throws InterruptedException {
        SubscribeToTopicRequest request = new SubscribeToTopicRequest().withTopic(topic);
        return clientV2.subscribeToTopic(request, this::onStreamEvent,
                        Optional.of(this::onStreamError),
                        Optional.of(this::onStreamClosed));
    }

    private void onStreamEvent(SubscriptionResponseMessage subscriptionResponseMessage) {
        try {
            BinaryMessage binaryMessage = subscriptionResponseMessage.getBinaryMessage();
            String message = new String(binaryMessage.getMessage(), StandardCharsets.UTF_8);
            String topic = binaryMessage.getContext().getTopic();
            System.out.printf("Received new message on topic %s: %s%n", topic, message);
        } catch (Exception e) {
            System.err.println("Exception occurred while processing subscription response " +
                    "message.");
            e.printStackTrace();
        }
    }

    private boolean onStreamError(Throwable error) {
        System.err.println("Received a stream error.");
        error.printStackTrace();
        return false; // Return true to close stream, false to keep stream open.
    }

    private void onStreamClosed() {
        System.out.println("Subscribe to topic stream closed.");
    }
}
